python/ray/_raylet.pyx

```diff
@@ -482,15 +629,34 @@ cdef class CoreWorker:
         return has_object
 
     def put_serialized_object(self, serialized_object, ObjectID object_id,
-                              int memcopy_threads=6):
+                              int memcopy_threads=6, c_bool put_async=False):
         cdef:
             shared_ptr[CBuffer] data
             shared_ptr[CBuffer] metadata
+            shared_ptr[CRayObject] ray_object
             CObjectID c_object_id = object_id.native()
-            size_t data_size
+            int64_t data_size
 
         data_size = serialized_object.total_bytes
 
+        # If we are returning values from a task, and the return objects are
+        # small, we can batch the puts until task batch completion. This saves
+        # a lot of IPCs to plasma for small objects.
+        if put_async and data_size < RayConfig.instance().size_limit():
+            data = dynamic_pointer_cast[
+                CBuffer, LocalMemoryBuffer](
+                    make_shared[LocalMemoryBuffer](
+                        <uint8_t*>NULL, data_size, True))
+            stream = pyarrow.FixedSizeBufferWriter(
+                pyarrow.py_buffer(Buffer.make(data)))
+            serialized_object.write_to(stream)
+            ray_object = make_shared[CRayObject](data, metadata)
+            with nogil:
+                check_status(self.core_worker.get().Objects().PutAsync(
```

Comment:
The implication with this is the objects put with `PutAsync` MUST not be consumed by `Get` with timeout as -1 in the same task, otherwise the code would deadlock. This looks a bit brittle ...

Reply:
It shouldn't be able to happen, since tasks are only batche if they have the same input dependencies. Though perhaps it is possible this happens via an object id passed indirectly, will have to think about this more.